samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java

```diff
@@ -383,6 +394,50 @@ public void onResourcesCompleted(List<SamzaResourceStatus> resourceStatuses) {
     }
   }
 
+  @Override
+  public void onStreamProcessorLaunchSuccess(SamzaResource resource) {
+
+    // 1. Obtain the Samza container Id for the pending container on this resource.
+    String containerId = getPendingContainerId(resource.getResourceID());
+    log.info("Successfully started container ID: {} on resource: {}", containerId, resource);
+
+    // 2. Remove the container from the pending buffer and add it to the running buffer. Additionally, update the
+    // job-health metric.
+    if (containerId != null) {
+      log.info("Moving containerID: {} on resource: {} from pending to running state", containerId, resource);
+      state.pendingContainers.remove(containerId);
+      state.runningContainers.put(containerId, resource);
+
+      if (state.neededContainers.decrementAndGet() == 0) {
+        state.jobHealthy.set(true);
+      }
+    } else {
+      log.warn("SamzaResource {} was not in pending state. Got an invalid callback for a launch request that " +
+          "was not issued", resource);
+    }
+  }
+
+  @Override
+  public void onStreamProcessorLaunchFailure(SamzaResource resource, Throwable t) {
+    log.error("Got a launch failure for SamzaResource {} with exception {}", resource, t);
+    // 1. Release resources for containers that failed back to YARN
+    log.info("Releasing unstartable container {}", resource.getResourceID());
+    clusterResourceManager.releaseResources(resource);
+
+    // 2. Obtain the Samza container Id for the pending container on this resource.
+    String containerId = getPendingContainerId(resource.getResourceID());
+    log.info("Failed container ID: {} for resourceId: {}", containerId, resource.getResourceID());
```

Comment:
Do we also need to remove the containerId from the state.pendingContainers?

Reply:
@xinyuiscool github isn't automatically expanding the changes in YarnClusterManager. Expand that diff and you should see the callback invocation.

Reply:
Thanks Jake! This behavior is annoying in github.